Scram uo

tombentley opened the pull request:
### Type of change

- Enhancement / new feature

### Description

This builds on #755 and adds support for specifying `KafkaUsers` with SCRAM-SHA-512 authentication.

* Generating the scram sha credentials is done in a subprocess of the UO using the Scala class corresponding to `kafka-configs.sh`. 
* Factored out a common `ProcessHelper` which is also used by the TO. 
* The tests for this require a running zookeeper, so I have refactored the `EmbeddedZooKeeper` fromthe TO tests into a new test artifact. I have also moved TestUtils there too.
* Added ST for the scram-sha authentication.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

scholzj commented on examples/kafka/kafka-ephemeral.yaml:
```diff
@@ -6,7 +6,9 @@ spec:
-      plain: {}
+      plain:
+        authentication:
+          type: scram-sha-512
```
Do we really want to do this? We haven't enabled TLS auth by default.

scholzj commented on examples/user/kafka-user.yaml:
```diff
@@ -6,7 +6,7 @@ metadata:
 spec:
   authentication:
-    type: tls
+    type: scram-sha-512
```
Why are we replacing it? Why do we think SASL should be used here?

scholzj commented on operator-common/src/main/java/io/strimzi/operator/common/process/ProcessHelper.java:
```diff
@@ -0,0 +1,90 @@
+    private static final Logger LOGGER = LogManager.getLogger(ProcessHelper.class);
+
+    public static ProcessResult executeSubprocess(List<String> verifyArgs) throws IOException, InterruptedException {
+        // We choose to run the reassignment as an external process because the Scala class:
```
I guess this is not anymore only for reassignments, so we should fix the comments.

scholzj commented on user-operator/src/main/resources/log4j2.properties:
```diff
@@ -5,7 +5,7 @@ appender.console.name = STDOUT
 appender.console.layout.pattern = %d{yyyy-MM-dd HH:mm:ss} %-5p %c{1}:%L - %m%n
 
-rootLogger.level = ${env:STRIMZI_LOG_LEVEL:-INFO}
+rootLogger.level = ${env:STRIMZI_LOG_LEVEL:-DEBUG}
```
Is this intentional? Or committed by mistake?

tombentley commented:
> A big question ... did you addressed in any way the problem with changing username and ACLs?

I think I can safely say the answer is no, but could you remind me what the problem was?

scholzj commented:
The problem is that the TLS user has principal `User:CN=jakub`. Where as I assume the SCRAM-SHA user will be just `User:jakub`. So ideally, we want to make sure there are no ACL rights for `CN=jakub` when SCRAM-SHA is used and vice versa.

That is at least the ideal case ... We might pretend that if the UserOperator manages the user `CN=jakub` there will be no other user `jakub` because they would have to share the same `KafkaUser` object. But its not ideal.

tombentley commented:
IMHO one `KafkaUser` should correspond to one use in Kafka. That would make intuitive sense. I don't really understand why Kafka uses `User:CN=jakub` for TLS-authenticated users and plain old `User:jakub` for SASL-authenticated users. I guess it's just a consequence of X509 naming.

For example, if I have a `KafkaUser` with TLS authentication and I switch to SCRAM-SHA, I don't expect that to change the principal name (because it's the same `KafkaUser`, just a different authentication mechanism).

Why do you think this is the less ideal of the two options?

scholzj commented:
Well, the problem is that this is how Kafka does it. The biggest problem is that sooner or later someone might come with the idea of using more fields in the subject then just the CN. And it can have a possible meaning for us as well - indicating a namespace for example etc. In that case the the username would be `CN=jakub,O=Something,OU=SomethingElse`. So it is not necessarily just the `CN=`. So the question is whether there is a way how to unify the usernames.

tombentley commented:
The existing reconciliation code is already broken in how it maps between plain names and distinguished names because `KafkaUserModel.decodeUserName()` is not the inverse function of `getUserName()`. 

I think we need to understand better the use cases where anyone would need to use anything more than `CN=%s`.

scholzj commented:
Why do you think its not the inverse function? It was supposed to be.

scholzj commented:
I think there are some potential usecase for the full DN based names. As I said you can use it to distinguish namespaces if we ever need it in the future. With some authorization types we can also use some of the fields as groups. And then sometimes you don't know what _crazy_ idea might the users come up with. So I'm not 100% convinced it will be future proof if we change how the principal is created and remove the `CN=`. But I might be wrong of course.

scholzj commented:
There should be always just one CN. Multiple CNs would mean invalid cert. And we currently don't support any other fields. So I think that method is fine for the current situation ... although not future proof obviously.

I don't know what the users will need ... its certificates ... so there will be always some smart security guy with some strange rules which might not improve security but need to be fulfilled. That said, we could also say that right now we should just ignore what might or might not be and solve the current issue.

tombentley commented:
Where do we stand on merging this?

tombentley commented:
Created #825

tombentley commented:
The commits from master were because this PR is to merge into the scram-listeners PR, but I've rebased them both now, so it should be clearer. The previous commits I've already addressed. I've opened an issue for the Principal/ACLs problem. I don't think anything has substantially changed since your previous review @scholzj, but it's ultimately up to you if you want to review it again.

ppatierno commented on systemtest/src/test/java/io/strimzi/systemtest/KafkaST.java:
```diff
@@ -340,11 +352,103 @@ public void testSendMessagesTlsAuthenticated() {
+
+    /**
+     * Test sending messages over tls transport using mutual tls auth
+     */
```
Wrong Javadoc

ppatierno commented on systemtest/src/test/java/io/strimzi/systemtest/KafkaST.java:
```diff
@@ -340,11 +352,103 @@ public void testSendMessagesTlsAuthenticated() {
+
+    /**
+     * Test sending messages over tls transport using mutual tls auth
+     */
```
Wrong Javadoc

ppatierno commented on user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java:
```diff
@@ -70,14 +73,19 @@ protected KafkaUserModel(String namespace, String name, Labels labels) {
      * @param userSecret    ubernetes secret with existing user certificate
```
update Javadoc with new `passwordGenerator` parameter

ppatierno commented on user-operator/src/main/java/io/strimzi/operator/user/operator/KafkaUserOperator.java:
```diff
@@ -64,10 +67,13 @@
     public KafkaUserOperator(Vertx vertx,
```
I know that some parameters were missing even before this PR but can we update the Javadoc now, wdyt?

ppatierno commented on user-operator/src/main/java/io/strimzi/operator/user/operator/KafkaUserOperator.java:
```diff
@@ -121,9 +130,12 @@ protected void delete(Reconciliation reconciliation, Handler<AsyncResult<Void>>
         String namespace = reconciliation.namespace();
         String user = reconciliation.name();
 
+        // TODO delete the user config
```
I guess you can delete this TODO
